src/server/server_family.cc

```diff
@@ -522,6 +516,247 @@ bool IsReplicatingNoOne(string_view host, string_view port) {
+    SaveDfsSingle(nullptr);
+
+    // Save shard files.
+    auto cb = [this](Transaction* t, EngineShard* shard) {
+      SaveDfsSingle(shard);
+      return OpStatus::OK;
+    };
+    trans_->ScheduleSingleHop(std::move(cb));
+  }
+
+  // Start saving a dfs file on shard
+  void SaveDfsSingle(EngineShard* shard) {
+    auto& [snapshot, filename] = snapshots_[shard ? shard->shard_id() : shard_set->size()];
+
+    SaveMode mode = shard == nullptr ? SaveMode::SUMMARY : SaveMode::SINGLE_SHARD;
+    auto scripts = shard == nullptr ? GetLuaScripts() : StringVec{};
+
+    if (auto err = snapshot->Start(mode, filename, scripts); err) {
+      shared_err_ = err;
+      snapshot.reset();
+      return;
+    }
+
+    if (mode == SaveMode::SINGLE_SHARD)
+      snapshot->StartInShard(shard);
+  }
+
+  // Save a single rdb file
+  void SaveRdb() {
+    auto& [snapshot, _] = snapshots_.front();
+
+    if (!full_path_.has_extension())
+      full_path_ += ".rdb";
+    if (!is_cloud_)
+      full_path_ += ".tmp";
+
+    if (auto err = snapshot->Start(SaveMode::RDB, full_path_, GetLuaScripts()); err) {
+      snapshot.reset();
+      return;
+    }
+
+    auto cb = [snapshot = snapshot.get()](Transaction* t, EngineShard* shard) {
+      snapshot->StartInShard(shard);
+      return OpStatus::OK;
+    };
+    trans_->ScheduleSingleHop(std::move(cb));
+  }
+
+  void UpdateSaveInfo() {
+    fs::path resulting_path = full_path_;
+    if (use_dfs_format_)
+      SetExtension("summary", ".dfs", &resulting_path);
+    else
+      resulting_path.replace_extension();  // remove .tmp
+
+    double seconds = double(absl::ToInt64Milliseconds(absl::Now() - start_time_)) / 1000;
+    LOG(INFO) << "Saving " << resulting_path << " finished after "
+              << strings::HumanReadableElapsedTime(seconds);
+
+    auto save_info = make_shared<LastSaveInfo>();
+    for (const auto& k_v : rdb_name_map_) {
+      save_info->freq_map.emplace_back(k_v);
+    }
+    save_info->save_time = absl::ToUnixSeconds(start_time_);
+    save_info->file_name = resulting_path.generic_string();
+    save_info->duration_sec = uint32_t(seconds);
+
+    lock_guard lk{*save_mu_};
+    last_save_info_->swap(save_info);  // swap - to deallocate the old version outstide of the lock.
+  }
+
+  GenericError InitResources() {
+    if (is_cloud_ && !aws_) {
+      *aws_ = make_unique<cloud::AWS>("s3");
+      if (auto ec = aws_->get()->Init(); ec) {
+        aws_->reset();
+        return {ec, "Couldn't initialize AWS"};
+      }
+    }
+
+    snapshots_.resize(use_dfs_format_ ? shard_set->size() + 1 : 1);
+    for (auto& [snapshot, _] : snapshots_)
+      snapshot = make_unique<RdbSnapshot>(fq_threadpool_, aws_->get());
+    return {};
+  }
+
+  // Remove .tmp extension or delete files in case of error
+  void FinalizeFileMovement(bool has_error) {
+    if (is_cloud_)
+      return;
+
+    for (const auto& [_, filename] : snapshots_) {
+      if (has_error)
+        filesystem::remove(filename);
```

Comment:
We didn't clean up before didn't we?

Reply:
If that's the case, and if there is an error, shouldn't we also clean up the leftover files from the cloud? I am referring to the early return above

Reply:
Yes, we didn't, I added it.

The cloud feature was already semi-finished. First, it broke with the new .tmp save mechanism (I fixed it too). Second, it has almost no support for error handling and other stuff - I don't know how to rename/delete/manage files with the limited aws client we currently have. Need to add this in the future

Reply:
Definitely! Could you plz add an issue for the AWS related cleanup such that we have it somewhere logged?

P.s. good catch with the clean up -- I reviewed the `.tmp` changes and I completely missed it.